File: src/main/java/ch/epfl/dias/calcite/adapter/pelago/PelagoTable.java
package ch.epfl.dias.calcite.adapter.pelago;

import ch.epfl.dias.calcite.adapter.pelago.types.PelagoTypeParser;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;

import org.apache.calcite.jdbc.JavaTypeFactoryImpl;
import org.apache.calcite.plan.RelOptTable;
//import org.apache.calcite.rel.RelDeviceType;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.RelReferentialConstraint;
import org.apache.calcite.rel.RelReferentialConstraintImpl;
import org.apache.calcite.rel.metadata.RelMdUtil;
import org.apache.calcite.rel.type.RelDataType;
import org.apache.calcite.rel.type.RelDataTypeFactory;
import org.apache.calcite.rel.type.RelProtoDataType;
import org.apache.calcite.rex.RexBuilder;
import org.apache.calcite.rex.RexLiteral;
import org.apache.calcite.schema.*;
import org.apache.calcite.schema.impl.AbstractTable;
import org.apache.calcite.sql.type.SqlTypeName;
import org.apache.calcite.util.ImmutableBitSet;
import org.apache.calcite.util.Pair;
import org.apache.calcite.util.Source;
import org.apache.calcite.util.mapping.IntPair;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Based on:
 * https://github.com/apache/calcite/blob/master/example/csv/src/main/java/org/apache/calcite/adapter/csv/CsvTable.java
 */
public class PelagoTable extends AbstractTable implements TranslatableTable {
    protected final RelProtoDataType    protoRowType;
    protected final RelDataType         rowType;
    protected final Source              source      ;
    protected final String              name        ;
//    protected RelDataType               rowType     ;
    protected Map<String, ?>            type        ;
    protected Map<String, ?>            plugin      ;
    protected Long                      linehint    ;
    protected List<Map<String, ?>>      constraints ;

    protected Map<String, Table>        knownTables ;

    protected Statistic                             stats = null;
    protected ImmutableMap<ImmutableBitSet, Double> dCnt  = null;
    protected ImmutableMap<ImmutableBitSet, Pair<Object, Object>> ranges  = null;
    // This map IS mutable, it should be updated periodically
//    protected Map<ImmutableBitSet, > dCnt  = null;

    private PelagoTable(Source source, RelProtoDataType protoRowType, Map<String, ?> plugin, long linehint, List<Map<String, ?>> constraints) {
        this.source         = source    ;
        this.type           = null      ;
        this.rowType        = null      ;
        this.linehint       = linehint  ;
        this.plugin         = plugin    ;
        this.name           = source.path();

        this.protoRowType   = protoRowType;

        if (constraints != null) {
            this.constraints = constraints;
        } else {
            this.constraints = new ArrayList<>();
        }
    }

    private PelagoTable(String name, RelDataType rowType, List<Map<String, ?>> constraints) {
        this.source         = null;
        this.name           = name;
        this.type           = null      ;
        this.linehint       = Long.MAX_VALUE;
        this.plugin         = Map.of("type", "intermediate");

        this.protoRowType   = null;
        this.rowType        = rowType;

        if (constraints != null) {
            this.constraints = constraints;
        } else {
            this.constraints = new ArrayList<>();
        }
    }

    private PelagoTable(Source source, Map<String, ?> type, Map<String, ?> plugin, long linehint, List<Map<String, ?>> constraints) {
        this.source     = source    ;
        this.type       = type      ;
        this.rowType    = null      ;
        this.linehint   = linehint  ;
        this.plugin     = plugin    ;
        this.name       = source.path();

        this.protoRowType = null;

        if (constraints != null) {
            this.constraints = constraints;
        } else {
            this.constraints = new ArrayList<>();
        }
    }

    public RelDataType getRowType(RelDataTypeFactory typeFactory) {
        if (rowType != null && typeFactory == null) return rowType;
        if (protoRowType == null && typeFactory == null) typeFactory = new JavaTypeFactoryImpl();

        if (protoRowType != null) return protoRowType.apply(typeFactory);

        try {
            return PelagoTypeParser.parseType(typeFactory, type);
        } catch (IOException e) {
            return null;
        }
    }

    private int getColumnIndex(String col){
        return getRowType(null).getField(col, false, true).getIndex();
    }

    public void overwriteKnownTables(Map<String, Table> t){
        knownTables = t;
    }

    protected void initStatistics() {
        double rc = linehint;
        final List<ImmutableBitSet> keys = Lists.newArrayList();
        ImmutableMap.Builder dCntBuilder = ImmutableMap.builder();
        ImmutableMap.Builder rangesBuilder = ImmutableMap.builder();
//	  final Content content = supplier.get();
//	  for (Ord<Column> ord : Ord.zip(content.columns)) {
//	    if (ord.e.cardinality == content.size) {
//	      keys.add(ImmutableBitSet.of(ord.i));
//	    }
//	  }
//        keys.add(ImmutableBitSet.of(0));

        ImmutableList.Builder<RelReferentialConstraint> constr = ImmutableList.builder();

        for (Map<String, ?> c: constraints){
            String type = ((String) c.get("type")).toLowerCase();
            switch (type) {
            case "primary_key":
            case "unique": {
                List<String> columns = ((List<String>) c.get("columns"));
                assert (columns.size() > 0);

                ImmutableBitSet.Builder k = ImmutableBitSet.builder();
                for (String col : columns) {
                    k.set(getColumnIndex(col));
                }
                keys.add(k.build());
                break;
            }
            case "distinct_cnt": {
                List<String> columns = ((List<String>) c.get("columns"));
                assert (columns.size() > 0);

                ImmutableBitSet.Builder k = ImmutableBitSet.builder();
                for (String col : columns) {
                    k.set(getColumnIndex(col));
                }
                dCntBuilder.put(k.build(), ((Number) c.get("values")).doubleValue());
                break;
            }
            case "range": {
                String column = ((String) c.get("column"));

                int index = getColumnIndex(column);
                ImmutableBitSet col = ImmutableBitSet.of(index);

                Object litmin = c.getOrDefault("min", null);
                Object litmax = c.getOrDefault("max", null);

                rangesBuilder.put(col, Pair.of(litmin, litmax));
                break;
            }
            case "foreign_key": {
                List<String> columns = ((List<String>) c.get("columns"));
                String tableName = knownTables.entrySet().stream().
                    filter(x -> x.getValue() == this).findAny().get().getKey();

                String ref = ((String) c.get("referencedTable"));

                ImmutableList.Builder<IntPair> refs = ImmutableList.builder();

                List<Map<String, String>> pairs = (List<Map<String, String>>) c.get("references");

                for (Map<String, String> p: pairs){
                    refs.add(IntPair.of(
                        getColumnIndex(p.get("referee")),
                        ((PelagoTable) knownTables.get(ref)).getColumnIndex(p.get("referred"))
                    ));
                }

                constr.add(
                    RelReferentialConstraintImpl.of(
                        ImmutableList.of("SSB", tableName),
                        ImmutableList.of("SSB", ref),
                        refs.build()
                    )
                );
                break;
            }
            default: {
                System.err.println("Unknown statistic: " + type);
                break;
            }
            }
        }

        dCnt = dCntBuilder.build();
        ranges = rangesBuilder.build();
        stats = Statistics.of(rc, keys, constr.build(), ImmutableList.of());
    }


    public Statistic getStatistic() {
        if (stats == null) initStatistics();
        return stats;
    }

    /** Returns an array of integers {0, ..., n - 1}. */
    private static int[] identityList(int n) {
        int[] ints = new int[n];
        for (int i = 0; i < n; i++) ints[i] = i;
        return ints;
    }

    public RelNode toRel(
            RelOptTable.ToRelContext context,
            RelOptTable relOptTable) {
        // Request all fields.
//        context.getCluster().getPlanner().addRelTraitDef(RelDistributionTraitDef.INSTANCE);
        final int fieldCount = relOptTable.getRowType().getFieldCount();
        final int[] fields = identityList(fieldCount);
        RelNode scan = PelagoTableScan.create(context.getCluster(), relOptTable, this, fields);
        if (getPacking() == RelPacking.Packed) scan = PelagoUnpack.create(scan, RelPacking.UnPckd);
        return scan;
    }

    public String getPelagoRelName(){
        return name;
    }

    public Map<String, ?> getPluginInfo(){
        return plugin;
    }

    public Long getLineHint(){
        return linehint;
    }

    public RelDeviceType   getDeviceType(){
        return RelDeviceType.X86_64;
    }

    public RelHomDistribution getHomDistribution(){
        return RelHomDistribution.SINGLE;
    }


    private static Long getLineHintFromPlugin(String name, Map<String, ?> plugin) throws MalformedPlugin {
        Object obj_linehint = plugin.getOrDefault("lines",  null);
        if (obj_linehint == null){
            obj_linehint = plugin.getOrDefault("linehint",  null);
        }

        Long linehint = null;
        if (obj_linehint != null) {
            if (obj_linehint instanceof Integer) {
                linehint = ((Integer) obj_linehint).longValue();
            } else if (obj_linehint instanceof Long){
                linehint = (Long) obj_linehint;
            } else {
                throw new MalformedPlugin("\"lines\" unrecognized type for \"lines\" during creation of " + name, name);
            }
        }

        if (linehint == null) {
            throw new MalformedPlugin("\"lines\" not found for table " + name, name);
        }

        return linehint;
    }

    public static PelagoTable create(Source source, String name, Map<String, ?> plugin, Map<String, ?> lineType  , List<Map<String, ?>> constraints) throws MalformedPlugin {
        return new PelagoTable(source, lineType, plugin, getLineHintFromPlugin(name, plugin), constraints);
    }

    public static PelagoTable create(Source source, String name, Map<String, ?> plugin, RelProtoDataType lineType) throws MalformedPlugin {
        return new PelagoTable(source, lineType, plugin, getLineHintFromPlugin(name, plugin), null);
    }

    public static PelagoTable create(String name, RelDataType lineType) throws MalformedPlugin {
        return new PelagoTable(name, lineType, null);
    }

    public RelPacking getPacking() {
        if (plugin.get("type").toString().equalsIgnoreCase("block")) return RelPacking.Packed;
        return RelPacking.UnPckd;
    }

    public Double getDistrinctValues(ImmutableBitSet cols) {
        if (dCnt == null) initStatistics();
        return dCnt.getOrDefault(cols, null);
    }

    public static BigInteger stringToNum(String x, int chars){
        BigInteger ret = BigInteger.valueOf(0);
        int len = Math.min(chars, x.length());
        for (int i = 0 ; i < len ; ++i) ret = ret.multiply(BigInteger.valueOf(256)).add(BigInteger.valueOf(Math.min(x.charAt(i), 255)));

        ret = ret.multiply(BigInteger.valueOf(256).pow(Math.max(x.length() - len, 0)));

        return ret;
    }

    public static double getPercentile(int min, Double max, Double val){
        return (val - min) / (max - min);
    }

    public static double getPercentile(BigDecimal min, BigDecimal max, BigDecimal val){
        return (val.subtract(min)).doubleValue() / (max.subtract(min)).doubleValue();
    }

    public static double getPercentile(BigInteger min, BigInteger max, BigInteger val){
        return (val.subtract(min)).doubleValue() / (max.subtract(min)).doubleValue();
    }

    public static double getPercentile(String start, String end, String q){
        if (q.compareTo(start) < 0) return 0;
        if (q.compareTo(end  ) > 0) return 1;
        if (end.equals(start)     ) return 1;

        int len = Math.max(Math.max(start.length(), end.length()), q.length());

        BigInteger max = stringToNum(end  , len);
        BigInteger min = stringToNum(start, len);
        BigInteger val = stringToNum(q    , len);

        return getPercentile(min, max, val);
    }

    public Double getPercentile(final ImmutableBitSet col, final RexLiteral val, final RexBuilder rexBuilder) {
        Double dist = getDistrinctValues(col);
        if (dist == null) return null;
        SqlTypeName type = getRowType(null).getFieldList().get(col.nextSetBit(0)).getType().getSqlTypeName();
        if (type == SqlTypeName.CHAR || type == SqlTypeName.VARCHAR || type == SqlTypeName.DECIMAL || type == SqlTypeName.BIGINT || type == SqlTypeName.INTEGER){
            Pair r = ranges.getOrDefault(col, null);
            if (r != null && r.left != null && r.right != null) {
                // While currently we only support VARCHAR literals for ranges, keep the extra step through RexLiteral here to allow for easier generalization
                if (type == SqlTypeName.CHAR || type == SqlTypeName.VARCHAR) {
                    RexLiteral rmin = (RexLiteral) rexBuilder.makeLiteral(r.left , val.getType(), true);
                    RexLiteral rmax = (RexLiteral) rexBuilder.makeLiteral(r.right, val.getType(), true);
                    String min = rmin.getValueAs(String.class);
                    String max = rmax.getValueAs(String.class);
                    return getPercentile(min, max, val.getValueAs(String.class));
                } else {
                    try {
                        BigDecimal min = new BigDecimal(r.left.toString());
                        BigDecimal max = new BigDecimal(r.right.toString());
                        return getPercentile(min, max, val.getValueAs(BigDecimal.class));
                    }catch (Exception e){
                        return null;
                    }
                }
            }
        }
        return RelMdUtil.numDistinctVals(dist, dist * 0.5)/dist;
    }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/schema/PelagoTable.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */
package ch.epfl.dias.calcite.adapter.pelago.schema

import ch.epfl.dias.calcite.adapter.pelago.traits.{
  RelDeviceType,
  RelHomDistribution,
  RelPacking
}
import ch.epfl.dias.calcite.adapter.pelago.types.PelagoTypeParser
import com.google.common.collect.{ImmutableList, ImmutableMap, Lists}
import org.apache.calcite.jdbc.JavaTypeFactoryImpl
import org.apache.calcite.plan.RelOptTable
import org.apache.calcite.rel.`type`.{
  RelDataType,
  RelDataTypeFactory,
  RelProtoDataType
}
import org.apache.calcite.rel.hint.RelHint
import org.apache.calcite.rel.logical.LogicalTableScan
import org.apache.calcite.rel.metadata.RelMdUtil
import org.apache.calcite.rel.{
  RelCollation,
  RelReferentialConstraint,
  RelReferentialConstraintImpl
}
import org.apache.calcite.rex.{RexBuilder, RexLiteral}
import org.apache.calcite.schema._
import org.apache.calcite.schema.impl.AbstractTable
import org.apache.calcite.sql.`type`.SqlTypeName
import org.apache.calcite.util.mapping.IntPair
import org.apache.calcite.util.{ImmutableBitSet, Pair, Source}

import java.io.IOException
import java.math.{BigDecimal, BigInteger}
import java.util.Objects
import java.{lang, util}
import scala.collection.JavaConverters._

/**
  * Based on:
  * https://github.com/apache/calcite/blob/master/example/csv/src/main/java/org/apache/calcite/adapter/csv/CsvTable.java
  */
object PelagoTable {
  @throws[MalformedPlugin]
  private def getLineHintFromPlugin(
      name: String,
      plugin: util.Map[String, Any]
  ) = {
    var obj_linehint = plugin.getOrDefault("lines", null)
    if (obj_linehint == null)
      obj_linehint = plugin.getOrDefault("linehint", null)
    var linehint: java.lang.Long = null
    if (obj_linehint != null)
      obj_linehint match {
        case integer: Integer => linehint = integer.longValue
        case l: Long          => linehint = l
        case _ =>
          throw new MalformedPlugin(
            "\"lines\" unrecognized type for \"lines\" during creation of " + name,
            name
          )
      }
    if (linehint == null)
      throw new MalformedPlugin("\"lines\" not found for table " + name, name)
    linehint
  }

  @throws[MalformedPlugin]
  def create(
      source: Source,
      name: String,
      plugin: util.Map[String, Any],
      lineType: util.Map[String, Any],
      constraints: util.List[util.Map[String, Any]]
  ) =
    new PelagoTable(
      source,
      lineType,
      plugin,
      getLineHintFromPlugin(name, plugin),
      constraints,
      name
    )

  @throws[MalformedPlugin]
  def create(
      source: Source,
      name: String,
      plugin: util.Map[String, Any],
      lineType: RelProtoDataType,
      alias: String
  ) =
    new PelagoTable(
      source,
      lineType,
      plugin,
      getLineHintFromPlugin(name, plugin),
      null,
      alias
    )

  @throws[MalformedPlugin]
  def create(
      source: Source,
      name: String,
      plugin: util.Map[String, Any],
      lineType: RelProtoDataType
  ): PelagoTable = create(source, name, plugin, lineType, name)

  @throws[MalformedPlugin]
  def create(name: String, lineType: RelDataType) =
    new PelagoTable(name, lineType, null)

  private def stringToNum(x: String, chars: Int) = {
    var ret = BigInteger.valueOf(0)
    val len = Math.min(chars, x.length)
    for (i <- 0 until len) {
      ret = ret
        .multiply(BigInteger.valueOf(256))
        .add(BigInteger.valueOf(Math.min(x.charAt(i), 255)))
    }
    ret = ret.multiply(BigInteger.valueOf(256).pow(Math.max(x.length - len, 0)))
    ret
  }

  private def getPercentile(min: Int, max: Double, v: Double) =
    (v - min) / (max - min)

  private def getPercentile(min: BigDecimal, max: BigDecimal, v: BigDecimal) =
    v.subtract(min).doubleValue / max.subtract(min).doubleValue

  private def getPercentile(min: BigInteger, max: BigInteger, v: BigInteger) =
    v.subtract(min).doubleValue / max.subtract(min).doubleValue

  def getPercentile(start: String, end: String, q: String): Double = {
    if (q.compareTo(start) < 0) return 0
    if (q.compareTo(end) > 0) return 1
    if (end == start) return 1
    val len = Math.max(Math.max(start.length, end.length), q.length)
    val max = stringToNum(end, len)
    val min = stringToNum(start, len)
    val v = stringToNum(q, len)
    getPercentile(min, max, v)
  }
}

class PelagoTable extends AbstractTable with TranslatableTable {
  final protected var protoRowType: RelProtoDataType = null
  final protected var rowType: RelDataType = null
  final protected var source: Source = null
  final protected var name: String = null
  final var alias: String = null
//    protected RelDataType               rowType     ;
  protected var `type`: util.Map[String, Any] = null
  var plugin: util.Map[String, Any] = null
  protected var linehint = 0L
  protected var constraints: util.List[util.Map[String, Any]] = null
  protected var knownTables: util.Map[String, Table] = null
  protected var stats: Statistic = null
  protected var dCnt: ImmutableMap[ImmutableBitSet, java.lang.Double] = null
  protected var ranges: ImmutableMap[ImmutableBitSet, Pair[Any, Any]] = null
  def this(
      source: Source,
      protoRowType: RelProtoDataType,
      plugin: util.Map[String, Any],
      linehint: Long,
      constraints: util.List[util.Map[String, Any]],
      alias: String
  ) {
    this()
    this.source = source
    this.`type` = null
    this.rowType = null
    this.linehint = linehint
    this.plugin = plugin
    this.name = source.path
    this.alias = alias
    this.protoRowType = protoRowType
    this.constraints = Objects.requireNonNullElseGet(
      constraints,
      () => Lists.newArrayList[util.Map[String, Any]]
    )
  }

  def this(
      name: String,
      rowType: RelDataType,
      constraints: util.List[util.Map[String, Any]]
  ) {
    this()
    this.source = null
    this.name = name
    this.alias = name
    this.`type` = null
    this.linehint = Long.MaxValue
    this.plugin = ImmutableMap.of("type", "intermediate")
    this.protoRowType = null
    this.rowType = rowType
    this.constraints = Objects.requireNonNullElseGet(
      constraints,
      () => Lists.newArrayList[util.Map[String, Any]]
    )
  }

  def this(
      source: Source,
      `type`: util.Map[String, Any],
      plugin: util.Map[String, Any],
      linehint: Long,
      constraints: util.List[util.Map[String, Any]],
      alias: String
  ) {
    this()
    this.source = source
    this.`type` = `type`
    this.rowType = null
    this.linehint = linehint
    this.plugin = plugin
    this.name = source.path
    this.alias = alias
    this.protoRowType = null
    this.constraints = Objects.requireNonNullElseGet(
      constraints,
      () => Lists.newArrayList[util.Map[String, Any]]
    )
  }

  override def getRowType(typeFactory: RelDataTypeFactory): RelDataType = {
    if (rowType != null && typeFactory == null) return rowType
    val tfact =
      if (protoRowType == null && typeFactory == null)
        new JavaTypeFactoryImpl
      else
        typeFactory

    if (protoRowType != null) return protoRowType.apply(tfact)

    try PelagoTypeParser.parseType(tfact, `type`)
    catch {
      case _: IOException =>
        null
    }
  }

  private def getColumnIndex(col: String) =
    getRowType(null).getField(col, false, true).getIndex

  def overwriteKnownTables(t: util.Map[String, Table]): Unit =
    knownTables = t

  private def initStatistics(): Unit = {
    val rc = linehint
    val keys = Lists.newArrayList[ImmutableBitSet]
    val dCntBuilder = ImmutableMap.builder[ImmutableBitSet, java.lang.Double]
    val rangesBuilder = ImmutableMap.builder[ImmutableBitSet, Pair[Any, Any]]
//	  final Content content = supplier.get();
//	  for (Ord<Column> ord : Ord.zip(content.columns)) {
//	    if (ord.e.cardinality == content.size) {
//	      keys.add(ImmutableBitSet.of(ord.i));
//	    }
//	  }
//        keys.add(ImmutableBitSet.of(0));
    val constr = ImmutableList.builder[RelReferentialConstraint]
    for (c <- constraints.asScala) {
      val `type` = c.get("type").asInstanceOf[String].toLowerCase
      `type` match {
        case "primary_key" =>
        case "unique" =>
          val columns = c.get("columns").asInstanceOf[util.List[String]]
          assert(columns.size > 0)
          val k = ImmutableBitSet.builder
          for (col <- columns.asScala) { k.set(getColumnIndex(col)) }
          keys.add(k.build)

        case "distinct_cnt" =>
          val columns = c.get("columns").asInstanceOf[util.List[String]]
          assert(columns.size > 0)
          val k = ImmutableBitSet.builder
          for (col <- columns.asScala) { k.set(getColumnIndex(col)) }
          dCntBuilder.put(
            k.build,
            c.get("values").asInstanceOf[Number].doubleValue
          )

        case "range" =>
          val column = c.get("column").asInstanceOf[String]
          val index = getColumnIndex(column)
          val col = ImmutableBitSet.of(index)
          val litmin = c.getOrDefault("min", null)
          val litmax = c.getOrDefault("max", null)
          rangesBuilder.put(col, Pair.of(litmin, litmax))

        case "foreign_key" =>
          val columns = c.get("columns").asInstanceOf[util.List[String]]
          val tableName = knownTables.entrySet.stream
            .filter((x: util.Map.Entry[String, Table]) => x.getValue eq this)
            .findAny
            .get
            .getKey
          val ref = c.get("referencedTable").asInstanceOf[String]
          val refs = ImmutableList.builder[IntPair]
          val pairs = c
            .get("references")
            .asInstanceOf[util.List[util.Map[String, String]]]

          for (p <- pairs.asScala) {
            refs.add(
              IntPair.of(
                getColumnIndex(p.get("referee")),
                knownTables
                  .get(ref)
                  .asInstanceOf[PelagoTable]
                  .getColumnIndex(p.get("referred"))
              )
            )
          }
          constr.add(
            RelReferentialConstraintImpl.of(
              ImmutableList.of("SSB", tableName),
              ImmutableList.of("SSB", ref),
              refs.build
            )
          )

        case _ =>
          System.err.println("Unknown statistic: " + `type`)

      }
    }
    dCnt = dCntBuilder.build
    ranges = rangesBuilder.build
    stats =
      Statistics.of(rc, keys, constr.build, ImmutableList.of[RelCollation])
  }

  override def getStatistic: Statistic = {
    if (stats == null) initStatistics()
    stats
  }

  override def toRel(
      context: RelOptTable.ToRelContext,
      relOptTable: RelOptTable
  ): LogicalTableScan =
    LogicalTableScan.create(
      context.getCluster,
      relOptTable,
      util.List.of[RelHint]
    )

  def getPelagoRelName: String = name

  def getPluginInfo: util.Map[String, Any] = plugin

  def getLineHint: java.lang.Long = linehint

  def getDeviceType: RelDeviceType = RelDeviceType.X86_64

  def getHomDistribution: RelHomDistribution = RelHomDistribution.SINGLE

  def getPacking: RelPacking = {
    if (plugin.get("type").toString.contains("block")) return RelPacking.Packed
    RelPacking.UnPckd
  }
  def getDistrinctValues(cols: ImmutableBitSet): lang.Double = {
    if (dCnt == null) initStatistics()
    dCnt.getOrDefault(cols, null)
  }

  def getRangeValues(cols: ImmutableBitSet): Pair[Any, Any] = {
    if (ranges == null) initStatistics()
    ranges.getOrDefault(cols, null)
  }

  def getPercentile(
      col: ImmutableBitSet,
      `val`: RexLiteral,
      rexBuilder: RexBuilder
  ): lang.Double = {
    val dist = getDistrinctValues(col)
    if (dist == null) return null
    val `type` = getRowType(null).getFieldList
      .get(col.nextSetBit(0))
      .getType
      .getSqlTypeName
    if (
      (`type` eq SqlTypeName.CHAR) || (`type` eq SqlTypeName.VARCHAR) || (`type` eq SqlTypeName.DECIMAL) || (`type` eq SqlTypeName.BIGINT) || (`type` eq SqlTypeName.INTEGER)
    ) {
      val r = ranges.getOrDefault(col, null)
      if (r != null && r.left != null && r.right != null) { // While currently we only support VARCHAR literals for ranges, keep the extra step through RexLiteral here to allow for easier generalization
        if ((`type` eq SqlTypeName.CHAR) || (`type` eq SqlTypeName.VARCHAR)) {
          val rmin = rexBuilder
            .makeLiteral(r.left, `val`.getType, true)
            .asInstanceOf[RexLiteral]
          val rmax = rexBuilder
            .makeLiteral(r.right, `val`.getType, true)
            .asInstanceOf[RexLiteral]
          val min = rmin.getValueAs(classOf[String])
          val max = rmax.getValueAs(classOf[String])
          return PelagoTable.getPercentile(
            min,
            max,
            `val`.getValueAs(classOf[String])
          )
        } else
          try {
            val min = new BigDecimal(r.left.toString)
            val max = new BigDecimal(r.right.toString)
            return PelagoTable.getPercentile(
              min,
              max,
              `val`.getValueAs(classOf[BigDecimal])
            )
          } catch {
            case e: Exception =>
              return null
          }
      }
    }
    RelMdUtil.numDistinctVals(dist, dist * 0.5) / dist
  }
}
